### PR TITLE
Add watcher

### PR DESCRIPTION
**Pull request by andreas131989**
Implementing watcher for deployment objects. It adds a validation label function to check the existence of the opt-in label.
The watcher is implemented as a separate controller and will only reconcile the object in case of the label exists and is set to true.
To do that, it watches all deployments across all namespaces by using the watch verb in the cluster role.

Once we start watching different resources, the structure of the watcher will also change a bit as a controller can watch only 1 primary resource. 
However for the current state of the operator, this controller is working as expected.

### CODE REVIEW

File: controllers/watcher.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2019 LitmusChaos Authors
```
Comment (alex-glv) [inline controllers/watcher.go:2]: Probably from copy/paste.

File: controllers/watcher.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	deployment := v1.Deployment{}
+	err := r.Client.Get(ctx, req.NamespacedName, &deployment)
+	if err != nil {
+		log.Error(err, "Failed to get the deployment data")
+		return ctrl.Result{}, err
+	}
+
+	// The first thing we need to do is determine if the deployment has the opt-in label and if it's set to true
+	// If neither of these conditions is met, then we won't reconcile.
+	optedinDeployment, err := validations.OptinLabelExists(deployment)
```
Comment (alex-glv) [inline controllers/watcher.go:68]: A small thing, but this variable name reads to me as "deployment that's opted in", however, it's a "deployment that has optin label"

File: controllers/watcher.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	// Next step after we are certain that we have an object to reconcile, we need to get the state definitions
+	stateDefinitions, err := states.GetClusterScalingStateDefinitions(ctx, r.Client)
+	if err != nil {
+		log.Error(err, "Failed to get ClusterStateDefinitions")
+		return ctrl.Result{}, err
+	}
+
+	// We need to calculate the desired state before we try to reconcile the deployment
+	finalState, err := reconciler.GetAppliedState(ctx, r.Client, req.Namespace, stateDefinitions, states.State{})
+	if err != nil {
```
Comment (alex-glv) [inline controllers/watcher.go:91]: If err != nil, is it going to cause issues when trying to reconcile? In this case, the error is just logged.

File: controllers/watcher.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+		log.Error(err, "Failed to get the deployment data")
+		return ctrl.Result{}, err
+	}
+
+	// The first thing we need to do is determine if the deployment has the opt-in label and if it's set to true
+	// If neither of these conditions is met, then we won't reconcile.
+	optedinDeployment, err := validations.OptinLabelExists(deployment)
+	if err != nil {
+		if strings.Contains(err.Error(), "Opt-in label was not found") {
```
Comment (alex-glv) [inline controllers/watcher.go:70]: The error text might change where it's generated, but this  text may stay here. Maybe it's better to move this error text to const?

**Review by alex-glv — CHANGES_REQUESTED** [review]
The only thing - comparing error contents, the rest is good!

File: controllers/watcher.go (hunk 1 above)
Comment (andreas131989) [reply controllers/watcher.go:2]: Ah right :D

File: controllers/watcher.go (hunk 3 above)
Comment (andreas131989) [reply controllers/watcher.go:90]: You mean that the return statement is missing, right? Yes, missed that , i'll add it

File: controllers/watcher.go (hunk 2 above)
Comment (andreas131989) [reply controllers/watcher.go:67]: I see your point, but actually this variable is the value of the opt-in label. The validation function essentially returns the value of the key label. So maybe a better name is optinLabel or something like that

**Review by alex-glv — APPROVED** [review]

File: controllers/watcher.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
+/*
```
Comment (chris-cmsoft) [inline controllers/watcher.go:1]: We can maybe just use a different name for the file. Something liek deployment_watch_controller just to keep it standard to the rest.

File: controllers/watcher.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+func (r *Watcher) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	log := r.Log.
+		WithValues("reconciler kind", "Watcher").
+		WithValues("reconciler namespace", req.Namespace).
+		WithValues("reconciler object", req.Name)
+
+	// Fetch the deployment data
+	deployment := v1.Deployment{}
+	err := r.Client.Get(ctx, req.NamespacedName, &deployment)
```
Comment (chris-cmsoft) [inline controllers/watcher.go:59]: Use only r.Get() here

File: controllers/watcher.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Failed to validate the opt-in label")
+		return ctrl.Result{}, err
+	}
+
+	if !optinLabel {
+		log.Info("Deployment opted out. No reconciliation")
+		return ctrl.Result{}, nil
+	}
```
Comment (chris-cmsoft) [inline controllers/watcher.go:79]: Not necessary now, but maybe later we can refactor this to a common method for func IsOptedInViaLabel(Deployment) (bool, error) {}

**Review by chris-cmsoft — APPROVED** [review]

**Pull request merged**